src/MonoTorrent.Client/MonoTorrent/TorrentFile.cs

```diff
@@ -109,40 +139,64 @@ internal static ITorrentFile[] Create (int pieceLength, params long[] lengths)
 
-        internal static ITorrentFile[] Create (int pieceLength, params (string path, long length, ReadOnlyMemory<byte> md5sum, ReadOnlyMemory<byte> ed2k, ReadOnlyMemory<byte> sha1)[] files)
+        internal static ITorrentFile[] Create (int pieceLength, TorrentFileTuple[] files)
         {
             long totalSize = 0;
+
+            // register padding file byte counts into padding field of the real predecessor file
+            for (int t = 0, real = -1; t < files.Length; t++) {
+                if ((files[t].attributes & TorrentFileAttributes.Padding) != 0) {
+                    if (real < 0) {
+                        // this will only happen if the first file is a padding file, bep-0047 doesn't seem to forbid that
+                        // Do not try to handle this, it would be very messy.
+                        throw new ArgumentException ("Can not handle torrent that starts with padding");
+                    } else {
+                        // add the count to it will also work in case of consecutive padding files, also slightly edge-case-y
+                        files[real].padding += files[t].length;
+                    }
+                } else {
+                    real = t;
+                }
+            }
+
+            // now we can forget the padding tuples
+            files = files.Where (f => (f.attributes & TorrentFileAttributes.Padding) == 0).ToArray ();
+
             var results = new List<ITorrentFile> (files.Length);
             for (int i = 0; i < files.Length; i++) {
+                // If a file is of length zero, it's should be treated as being part of the first piece.
+                if (files[i].length == 0) {
+                    results.Add (new TorrentFile (files[i].path!, 0, 0, 0, 0, files[i].attributes, 0));
+                    continue;
+                }
+
                 var length = files[i].length;
+                var padding = files[i].padding;
 
                 var pieceStart = (int) (totalSize / pieceLength);
-                var pieceEnd = (int) ((totalSize + length) / pieceLength);
+                var pieceEnd = length > 0 ? (int) ((totalSize + length - 1) / pieceLength) : pieceStart;
+                var pieceEndWithPadding = (length + padding) > 0 ? (int) ((totalSize + (length + padding) - 1) / pieceLength) : pieceStart;
                 var startOffsetInTorrent = totalSize;
-                if ((totalSize + length) % pieceLength == 0)
-                    pieceEnd--;
-
-                if (length == 0) {
-                    pieceStart = i > 0 ? results[i - 1].StartPieceIndex : 0;
-                    pieceEnd = i > 0 ? results[i - 1].StartPieceIndex : 0;
-                    startOffsetInTorrent = i > 0 ? results[i - 1].OffsetInTorrent : 0;
+                // catch pathological case of too much padding
+                if (pieceEnd != pieceEndWithPadding) {
+                    throw new ArgumentException ("A file in the torrent has more padding than needed.");
```

Comment:
This check was helpful as it triggered when I created the testcase to see what happened when `file.Length == PieceLength` :)